cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -0,0 +1,342 @@
+import Distribution.Types.PackageDescription
+         ( PackageDescription(..), emptyPackageDescription )
+import Distribution.Types.PackageName.Magic
+         ( fakePackageId )
+import Language.Haskell.Extension
+         ( Language(..) )
+import Distribution.Client.HashValue
+         ( hashValue, showHashValue )
+import Distribution.Simple.Utils
+         ( readUTF8File )
+
+import Control.Concurrent.MVar
+         ( newEmptyMVar, putMVar, tryTakeMVar )
+import Control.Exception
+         ( bracket )
+import qualified Data.ByteString.Char8 as BS
+import Data.ByteString.Lazy ()
+import qualified Text.Parsec as P
+import System.Directory
+         ( getTemporaryDirectory, removeDirectoryRecursive, doesFileExist, canonicalizePath )
+import System.FilePath
+         ( (</>), takeExtension )
+
+
+-- | Get the directory where script builds are cached.
+--
+-- CABAL_DIR\/script-builds\/
+getScriptCacheDirectoryRoot :: IO FilePath
+getScriptCacheDirectoryRoot = do
+  cabalDir <- getCabalDir
+  return $ cabalDir </> "script-builds"
+
+-- | Get the hash of a script path.
+getScriptHash :: String -> FilePath -> IO String
+getScriptHash prefix script = showHashValue . hashValue . fromString . (prefix ++) <$> canonicalizePath script
+
+-- | Get the directory for caching a script build.
+--
+-- The only identity of a script is it's absolute path, so append the
+-- hashed path to CABAL_DIR\/script-builds\/ to get the cache directory.
+getScriptCacheDirectory :: String -> FilePath -> IO FilePath
+getScriptCacheDirectory prefix script = (</>) <$> getScriptCacheDirectoryRoot <*> getScriptHash prefix script
+
+-- | Get the directory for caching a script build and ensure it exists.
+--
+-- The only identity of a script is it's absolute path, so append the
+-- hashed path to CABAL_DIR\/script-builds\/ to get the cache directory.
+ensureScriptCacheDirectory :: Verbosity -> String -> FilePath -> IO FilePath
+ensureScriptCacheDirectory verbosity prefix script = do
+  cacheDir <- getScriptCacheDirectory prefix script
+  createDirectoryIfMissingVerbose verbosity True cacheDir
+  return cacheDir
+
+-- | What your command should do when no targets are found.
+data AcceptNoTargets
+  = RejectNoTargets -- ^ die on 'TargetSelectorNoTargetsInProject'
+  | AcceptNoTargets -- ^ return a default 'TargetSelector'
+  deriving (Eq, Show)
+
+-- | Information about the context in which we found the 'TargetSelector's.
+data TargetContext
+  = ProjectContext -- ^ The target selectors are part of a project.
+  | GlobalContext  -- ^ The target selectors are from the global context.
+  | ScriptContext { scriptPath :: FilePath, scriptExecutable :: Executable, scriptContents :: BS.ByteString }
+  -- ^ The target selectors refer to a script.
+  deriving (Eq, Show)
+
+-- | Determine whether the targets represent regular targets or a script
+-- and return the proper context and target selectors.
+-- Report problems if selectors are valid as neither regular targets or as a script.
```

Comment:
How are problems reported?
I suspect just like before, which was `print to stderr >> die`, but can we document that here as well?